Tasks/Gradle/sonarqube-common/sonarqube-rest.ts

```diff
@@ -0,0 +1,193 @@
+/// <reference path="../../../definitions/vsts-task-lib.d.ts" />
+
+import Q = require('q');
+import url = require('url');
+import http = require('http');
+import {IncomingMessage} from 'http';
+import https = require('https');
+
+import tl = require('../testlib/tasklib-wrapper');
+
+import {TaskReport} from  './taskreport';
+import {RestResponse} from  './restresponse';
+import {SonarQubeEndpoint} from './sonarqube-common';
+import sqCommon = require('./sonarqube-common');
+
+// Repeated query the server to determine if the task has finished.
+// Returns true if the task is finished, or false if the timeout was exceeded without a positive answer from the server.
+// Timeout and delay are in seconds, defaulting to 60 and 1 respectively if not specified
+export function waitForAnalysisCompletion(taskReport:TaskReport, timeout?:number, delay?:number):Q.Promise<boolean> {
+    var defer = Q.defer<boolean>();
+
+    // Default values
+    var timeout = timeout || 60;
+    var delay = delay || 1;
+
+    var isDone:boolean = false;
+    // Every [delay] seconds, call isAnalysisComplete()
+    var intervalObject = setInterval((taskReport) => {
+        isTaskComplete(taskReport)
+            .then((isDone:boolean) => {
+                if (isDone) {
+                    clearInterval(intervalObject);
+                    defer.resolve(true);
+                }
+            })
+            .fail((error) => {
+                // If anything goes wrong, delete the repeating call and reject.
+                clearInterval(intervalObject);
+                defer.reject(error);
+            })
+    }, delay * 1000, taskReport);
+
+    // After [timeout] seconds, delete the repeating call and return false.
+    setTimeout((intervalObject) => {
+        clearInterval(intervalObject);
+        defer.resolve(false);
+    }, timeout * 1000, intervalObject);
+
+    return defer.promise;
+}
+
+// Queries the server to determine if the task has finished, i.e. if the quality gate has been evaluated
+export function isTaskComplete(taskReport:TaskReport):Q.Promise<boolean> {
+    return getTaskDetails(taskReport)
+        .then((responseJson:any) => {
+            var taskStatus:string = responseJson.task.status;
+            return (taskStatus.toUpperCase() == 'SUCCESS');
+        });
+}
+
+// Query the server to determine the analysis id associated with the current task
+export function getAnalysisId(taskReport:TaskReport):Q.Promise<string> {
```

Comment:
There too many public functions related to this. All you need is the quality gate status. Why do you need to expose so many implementation details such as analysis id, analysis status etc.?